src/main/scala/cognite/spark/v1/wdl/WellDataLayerRelation.scala

```diff
@@ -0,0 +1,83 @@
+package cognite.spark.v1.wdl
+
+import cats.effect.IO
+import cognite.spark.v1.{CdfRelation, CdfSparkException, RelationConfig, WritableRelation}
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.sources.TableScan
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{Row, SQLContext}
+
+class WellDataLayerRelation(
+    config: RelationConfig,
+    model: String
+)(override val sqlContext: SQLContext)
+    extends CdfRelation(config, "well-data-layer")
+    with WritableRelation
+    with TableScan
+    with Serializable {
+
+  import cognite.spark.v1.CdpConnector._
+
+  override def schema: StructType =
+    cachedSchema.getOrElse(
+      throw new CdfSparkException("Failed to decode well-data-layer schema into StructType"))
+
+  private lazy val cachedSchema: Option[StructType] = getSchema
+
+  private def getSchema: Option[StructType] = {
+    val schemaAsString = client.wdl.getSchema(model).unsafeRunSync()
+    DataType.fromJson(schemaAsString) match {
+      case s @ StructType(_) => Some(s)
+      case _ => None
+    }
+  }
+
+  override def insert(rows: Seq[Row]): IO[Unit] = upsert(rows)
+
+  override def upsert(rows: Seq[Row]): IO[Unit] = {
+    val jsonObjects = rows.map(row => {
+      val rowJson = RowToJson.toJsonObject(row, schema)
+      if (rowJson.isEmpty) {
+        sys.error("Upsert invalid empty row!")
+      }
+      rowJson
+    })
+
+    client.wdl.setItems(getWriteUrlPart(model), jsonObjects)
+  }
+
+  // scalastyle:off cyclomatic.complexity
+  private def getWriteUrlPart(modelType: String): String =
+    modelType.replace("Ingestion", "") match {
+      case "Well" => "wells"
+      case "WellSource" => "wells"
+      case "Wellbore" => "wellbores"
+      case "WellboreSource" => "wellbores"
+      case "DepthMeasurement" => "measurements/depth"
+      case "TimeMeasurement" => "measurements/time"
+      case "RigOperation" => "rigoperations"
+      case "HoleSectionGroup" => "holesections"
+      case "WellTopGroup" => "welltops"
+      case "Npt" => "npt"
+      case "Nds" => "npt"
+      case "CasingSchematic" => "casings"
+      case "Trajectory" => "trajectories"
+      case "Source" => "sources"
+      case _ => sys.error(s"Unknown model type: $modelType")
```

Comment:
see comment above, may be use a local map instead of a long list of pattern matching 😁